### ISSUE TITLE
DNS challenge error when FQDN is a CNAME

### ISSUE THREAD

**Issue by happymcplaksin** [xenolf/lego#92]
Thanks for lego, it's awesome!

When I use the DNS challenge to request a cert for a FQDN that's a CNAME, I get the following. Assume iamacname.example.com is a CNAME for www.example.com

```
2016/02/01 13:06:25 [INFO][iamacname.example.com] acme: Obtaining bundled SAN certificate

2016/02/01 13:06:25 [INFO][iamacname.example.com] acme: Trying to solve DNS-01
2016/02/01 13:06:25 [INFO] acme: Please create the following TXT record in your DNS zone:
2016/02/01 13:06:25 [INFO] acme: _acme-challenge.iamacname.example.com. 120 IN TXT "U4BJAr7Mg1x2BSdYWeCDLvWSYPkj14XR3uohxan1_RU"
2016/02/01 13:06:25 [INFO] acme: Press 'Enter' when you are done
2016/02/01 13:06:25 [INFO] acme: You can now remove this TXT record from your DNS zone:
2016/02/01 13:06:25 [INFO] acme: _acme-challenge.iamacname.example.com. 120 IN TXT "..."
panic: interface conversion: dns.RR is *dns.CNAME, not *dns.SOA

goroutine 1 [running]:
github.com/xenolf/lego/acme.checkDNS(0x7ffcb3f28514, 0x13, 0xc820424690, 0x24, 0xc820424690)
        /home/happy/x/src/github.com/xenolf/lego/acme/dns_challenge.go:82 +0x355
github.com/xenolf/lego/acme.(*dnsChallenge).Solve(0xc82011c240, 0x0, 0x0, 0xc820432180, 0x6, 0xc820432190, 0x7, 0xc8203500e0, 0x68, 0xc8204240c0, ...)
        /home/happy/x/src/github.com/xenolf/lego/acme/dns_challenge.go:65 +0x5a6
github.com/xenolf/lego/acme.(*Client).solveChallenges(0xc820126080, 0xc82000a300, 0x1, 0x1, 0xc82000a300)
        /home/happy/x/src/github.com/xenolf/lego/acme/client.go:369 +0x250
github.com/xenolf/lego/acme.(*Client).ObtainCertificate(0xc820126080, 0xc820013990, 0x1, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/happy/x/src/github.com/xenolf/lego/acme/client.go:227 +0x25b
main.run(0xc8200dc240)
        /home/happy/x/src/github.com/xenolf/lego/cli_handlers.go:174 +0x173f
github.com/codegangsta/cli.Command.Run(0x9412b0, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x9dbd80, 0x3a, 0x0, ...)
        /home/happy/x/src/github.com/codegangsta/cli/command.go:163 +0x131d
github.com/codegangsta/cli.(*App).Run(0xc82008fe60, 0xc82000a0c0, 0xc, 0xc, 0x0, 0x0)
        /home/happy/x/src/github.com/codegangsta/cli/app.go:179 +0x1169
main.main()
        /home/happy/x/src/github.com/xenolf/lego/cli.go:126 +0x106f

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
        /usr/lib/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 9 [IO wait]:
net.runtime_pollWait(0x7f8f324ee000, 0x72, 0xc820012270)
        /usr/lib/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc820128060, 0x72, 0x0, 0x0)
        /usr/lib/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc820128060, 0x0, 0x0)
        /usr/lib/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc820128000, 0xc82031c000, 0x800, 0x800, 0x0, 0x7f8f34652050, 0xc820012270)
        /usr/lib/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc820030080, 0xc82031c000, 0x800, 0x800, 0x0, 0x0, 0x0)
        /usr/lib/go/src/net/net.go:172 +0xe4
crypto/tls.(*block).readFromUntil(0xc8201123f0, 0x7f8f34656b50, 0xc820030080, 0x5, 0x0, 0x0)
        /usr/lib/go/src/crypto/tls/conn.go:455 +0xcc
crypto/tls.(*Conn).readRecord(0xc8200d4840, 0xa14117, 0x0, 0x0)
        /usr/lib/go/src/crypto/tls/conn.go:540 +0x2d1
crypto/tls.(*Conn).Read(0xc8200d4840, 0xc820188000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /usr/lib/go/src/crypto/tls/conn.go:901 +0x167
net/http.noteEOFReader.Read(0x7f8f34663308, 0xc8200d4840, 0xc8200e67e8, 0xc820188000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /usr/lib/go/src/net/http/transport.go:1370 +0x67
net/http.(*noteEOFReader).Read(0xc82044a700, 0xc820188000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        <autogenerated>:126 +0xd0
bufio.(*Reader).fill(0xc8203ecfc0)
        /usr/lib/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc8203ecfc0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/lib/go/src/bufio/bufio.go:132 +0xcc
net/http.(*persistConn).readLoop(0xc8200e6790)
        /usr/lib/go/src/net/http/transport.go:876 +0xf7
created by net/http.(*Transport).dialConn
        /usr/lib/go/src/net/http/transport.go:685 +0xc78

goroutine 10 [select]:
net/http.(*persistConn).writeLoop(0xc8200e6790)
        /usr/lib/go/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
        /usr/lib/go/src/net/http/transport.go:686 +0xc9d
```

**Comment by xenolf** [comment]
Hey and thanks for your report!

Could you give me some background on your DNS setup?
The issue you are seeing is because a SOA query against `google-public-dns-a.google.com` returns a CNAME response for your FQDN instead of a SOA response (which imo is rather odd).

**Comment by happymcplaksin** [comment]
Hmm, maybe it's because I "cheated"? That is, our DNS looks like this:

We have a zone example.com and www.example.com is an A record inside that zone. Then we have iamacname.example.com as a CNAME for www.example.com and we're trying get a cert for iamacname.example.com.

Let's Encrypt wants the TXT record to be _acme-challenge.iamacname.example.com. Instead of creating a new zone _acme-challenge.iamacname.example.com and making the TXT record same-as-zone, I just made a TXT record in example.com named _acme-challenge.iamacname.

But even with this setup, I can't construct a dig command that returns a CNAME when asking for the SOA record. For example, this returns an SOA and not a CNAME:

dig -t soa iamacname.example.com @google-public-dns-a.google.com

**Comment by happymcplaksin** [comment]
Oh and I just noticed @mholt's cname-fix branch. With that I get the same "panic: interface conversion: dns.RR is *dns.CNAME, not *dns.SOA" error on line 82.

**Comment by mholt** [comment]
Huh? That's impossible; that type assertion doesn't exist on line 82 in the branch.

**Comment by happymcplaksin** [comment]
Ahem. It's my first time building Go from a repo so I must have made a mistake :) I'll try harder.

**Comment by mholt** [comment]
Ah, didn't know that. First make absolutely sure you're on the branch, then try clearing your $GOPATH/pkg folder and try again.

**Comment by happymcplaksin** [comment]
I continue to get an error on line 82. Tell me what to do step-by-step to make sure I'm really using the code in the branch? I know Git well enough to be certain I'm on the branch. But no matter how I 'go get' and 'go build" I still get the line 82 error.

Apologies for not being more of a GOer :)

**Comment by mholt** [comment]
No problem. You've already run `go get` so all you have to do is `cd` to $GOPATH/src/github.com/xenolf/lego and checkout the branch. Then `go build` will produce a binary you can run, `./lego`.

**Comment by happymcplaksin** [comment]
Oh, hmm. Then I think I did it right the first time and really do get the same error on line 82. Here's what I did just now to reproduce:

```
$ git checkout master
$ ./lego -x http-01 -x tls-sni-01 --dns manual -m ssladmin@example.com -d iamacname.example.com run > 1
$ echo | ./lego -x http-01 -x tls-sni-01 --dns manual -m ssladmin@example.com -d iamacname.example.com run > 1 2>&1
$ rm -rf pkg/
$ git checkout cname-fix
$ go build
$ echo | ./lego -x http-01 -x tls-sni-01 --dns manual -m ssladmin@example.com -d iamacname.example.com run > 2 2>&1
```

Output 1: https://gist.github.com/happymcplaksin/7d2b0fdd3e90c694dc82

Output 2: https://gist.github.com/happymcplaksin/e296ee228b2ced7326d4

I'm happy to be doing something wrong :) I just don't see the problem.

**Comment by mholt** [comment]
The `rm -rf` line should read `rm -rf $GOPATH/pkg`, but I'm still not sure if that'll fix it...

**Comment by happymcplaksin** [comment]
I had GOPATH set to (the full path of) the directory I was in so that `rm -rf` line was effectively `rm -rf $GOPATH/pkg`

Are you able to test with a CNAME to help determine if I'm just crazy? :)

**Comment by xenolf** [comment]
I was able to test it with a CNAME using Cloudflare and it worked fine for me.

My setup was CNAME www.example.com points to example.com.

**Comment by happymcplaksin** [comment]
Sorry for the delay; I went to re-test and it looks like the cname-fix branche is gone? I don't see the CNAME fix in master either.

**Comment by xenolf** [comment]
There is no fix as of yet. I tested it with the code as is and it worked for me.

**Comment by happymcplaksin** [comment]
Sure, I understand there's no final fix yet. I was going to re-test since I failed to make it work (presumably my fault). But the cname-fix branch disappeared. So I wonder if that's significant and I should wait for an alternate fix before retrying?

**Comment by mholt** [comment]
My fix wasn't correct, since Sebastian and I realized we don't understand the problem, because we can't reproduce it. We'd like to know your DNS setup, specifically what your zone file looks like.

**Comment by janeczku** [comment]
@happymcplaksin PR #96 will be more resilient to handle edge cases in DNS delegation. Maybe you give it a try?

```BASH
cd $GOPATH/src/github.com/xenolf/lego 
git fetch origin pull/96/head:dns-check-new
git checkout dns-check-new
go install
```

**Comment by happymcplaksin** [comment]
@janeczku That works! Thanks :)

FYI while testing I discovered an error on my part on our DNS. When I first added the TXT record for iamacname.example.com I used a new-ish script which I thought failed to add the TXT record entirely. So I added future TXT records by hand.

When testing today I discovered that the script had actually added the TXT record as _acme-challenge.example.com. This probably explains why the previous fix worked for @xenolf but not for me.

Today I tested PR #96 both with and without the oops TXT record and it worked in both cases.

Yay!

Thanks so much to all of you for the assistance and for lego!!

**Comment by happymcplaksin** [comment]
PS: I re-tested w/o the patch and w/o the oops TXT record and still get a failure on line 82. So PR #96 definitely helps.

**Comment by xenolf** [comment]
I will keep this issue open until #96 lands.

**Comment by xenolf** [comment]
Closed via #116
